projects/l2_pooling/noise_tolerance.py

```diff
@@ -0,0 +1,809 @@
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+from collections import defaultdict
+import math
+import random
+import os
+import time
+
+import matplotlib.pyplot as plt
+
+from htmresearch.algorithms.column_pooler import ColumnPooler
+from htmresearch.frameworks.layers.sensor_placement import greedySensorPositions
+
+L4_CELL_COUNT = 8*1024
+
+
+def createRandomObjectDescriptions(numObjects,
+                                   numLocationsPerObject,
+                                   featurePool=("A", "B", "C")):
+  """
+  Returns {"Object 1": [(0, "C"), (1, "B"), (2, "C"), ...],
+           "Object 2": [(0, "C"), (1, "A"), (2, "B"), ...]}
+  """
+  return dict(("Object %d" % i,
+               zip(xrange(numLocationsPerObject),
+                   [random.choice(featurePool)
+                    for _ in xrange(numLocationsPerObject)]))
+              for i in xrange(1, numObjects + 1))
+
+
+def noisy(pattern, noiseLevel, totalNumCells):
+  n = int(noiseLevel * 40)
+
+  noised = set(pattern)
+
+  noised.difference_update(random.sample(noised, n))
+
+  for _ in xrange(n):
+    while True:
+      v = random.randint(0, totalNumCells - 1)
+      if v not in pattern and v not in noised:
+        noised.add(v)
+        break
+
+  return noised
+
+
+def doExperiment(numColumns, l2Overrides, objectDescriptions, noiseFn,
+                 numInitialTraversals):
+
+  # For each column, keep a mapping from feature-location names to their SDRs
+  layer4sdr = lambda : set(random.sample(xrange(L4_CELL_COUNT), 40))
+  featureLocationSDRs = [defaultdict(layer4sdr) for _ in xrange(numColumns)]
+
+  poolers = [ColumnPooler(inputWidth=L4_CELL_COUNT,
+                          lateralInputWidths=[4096]*(numColumns-1),
+                          **l2Overrides)
+             for _ in xrange(numColumns)]
+
+  # Learn the objects
+  objectL2Representations = {}
+  for objectName, featureLocations in  objectDescriptions.iteritems():
+    for featureLocationName in featureLocations:
+      for _ in xrange(10):
+        allLateralInputs = [pooler.getActiveCells() for pooler in poolers]
+        for columnNumber, pooler in enumerate(poolers):
+          feedforwardInput = featureLocationSDRs[columnNumber][featureLocationName]
+          lateralInputs = [lateralInput
+                           for i, lateralInput in enumerate(allLateralInputs)
+                           if i != columnNumber]
+          pooler.compute(feedforwardInput, lateralInputs, learn=True)
```

Comment:
This is testing L2 in isolation, so the full-blown L2L4 setup isn't needed. And this is superior, since the lateralInputs are properly delayed by 1 timestep.